hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -1269,512 +1315,350 @@ object PruneDeadFields {
     ir: IR,
     requestedType: Type,
     memo: ComputeMutableState,
-  ): BindingEnv[BoxedArrayBuilder[Type]] = {
+    env: BindingEnv[TypeState] = BindingEnv.empty.createAgg.createScan,
+  ): Unit = {
+    def recurMax(ir: IR, childIdx: Int): IndexedSeq[TypeState] =
+      recur(ir, childIdx, ir.getChild(childIdx).asInstanceOf[IR].typ)
+
+    def recurMin(ir: IR, childIdx: Int): IndexedSeq[TypeState] =
+      recur(ir, childIdx, minimal(ir.getChild(childIdx).asInstanceOf[IR].typ))
+
+    def recur(ir: IR, childIdx: Int, requestedType: Type): IndexedSeq[TypeState] =
+      memoizeValueIR(ctx, ir, childIdx, requestedType, memo, env)
+
+    def recurMapMax(ir: IR, childIdx: Int, bindingsMap: mutable.Map[String, TypeState]): Unit =
+      recurMap(ir, childIdx, ir.getChild(childIdx).asInstanceOf[IR].typ, bindingsMap)
+
+    def recurMap(
+      ir: IR,
+      childIdx: Int,
+      requestedType: Type,
+      bindingsMap: mutable.Map[String, TypeState],
+    ): Unit = {
+      val bindings = Bindings.get(ir, childIdx).map { (name, typ) =>
+        val state = bindingsMap.getOrElseUpdate(name, TypeState(typ))
+        assert(typ == state.origType)
+        state
+      }
+      memoizeValueIR(
+        ctx,
+        ir.getChild(childIdx).asInstanceOf[IR],
+        requestedType,
+        memo,
+        env.extend(bindings),
+      )
+    }
```

Comment:
following the chain of delegation is quite hard here and I've never liked the pattern of passing around child-indexes everywhere. I think it would be clearer if you threaded the bindings through instead of the parent ir and the child index.

Reply:
Completely agree, and this is one of the things I'm working towards, trying to simplify the binding structure and how it's represented. But for now, this is the only way I could find to make PruneDeadFields delegate to Bindings instead of redundantly encoding the same information, which will help me make those changes.

Reply:
absolutely, not suggesting you implement this now. just lamenting about the proliferation of indices.